### PR TITLE
Fix Vector'Diagonal (and Transpose as well) to avoid infinite recursion.

### LINKED ISSUE #517: Stack overflow and segfault for x' * D * x
```jl

julia> using LinearAlgebra

julia> x = [1, 2, 3]
3-element Array{Int64,1}:
 1
 2
 3

julia> D = Diagonal(x)
3×3 Diagonal{Int64,Array{Int64,1}}:
  1  ⋅  ⋅
 ⋅  2  ⋅
 ⋅  ⋅  3

julia> x' * D * x
ERROR: StackOverflowError:
Stacktrace:
 [1] Type at ./boot.jl:391 [inlined]
 [2] similar at ./array.jl:262 [inlined]
 [3] similar at /Users/davidsanders/development/julia-dev/usr/share/julia/stdlib/v0.7/LinearAlgebra/src/adjtrans.jl:143 [inlined]
 [4] copymutable at ./abstractarray.jl:798 [inlined]
 [5] copy at ./abstractarray.jl:748 [inlined]
 [6] *(::Adjoint{Int64,Array{Int64,1}}, ::Diagonal{Int64,Array{Int64,1}}) at /Users/davidsanders/development/julia-dev/usr/share/julia/stdlib/v0.7/LinearAlgebra/src/diagonal.jl:484 (repeats 79998 times)

julia> x' * Diagonal(x) * x
Segmentation fault: 11
```

```
julia> versioninfo()
Julia Version 0.7.0-DEV.4909
Commit 5912b14 (2018-04-19 21:08 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin15.0.0)
  CPU: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
Environment:
```

### LINKED ISSUE #26863: Stack overflow and segfault for x' * D * x
```jl

julia> using LinearAlgebra

julia> x = [1, 2, 3]
3-element Array{Int64,1}:
 1
 2
 3

julia> D = Diagonal(x)
3×3 Diagonal{Int64,Array{Int64,1}}:
  1  ⋅  ⋅
 ⋅  2  ⋅
 ⋅  ⋅  3

julia> x' * D * x
ERROR: StackOverflowError:
Stacktrace:
 [1] Type at ./boot.jl:391 [inlined]
 [2] similar at ./array.jl:262 [inlined]
 [3] similar at /Users/davidsanders/development/julia-dev/usr/share/julia/stdlib/v0.7/LinearAlgebra/src/adjtrans.jl:143 [inlined]
 [4] copymutable at ./abstractarray.jl:798 [inlined]
 [5] copy at ./abstractarray.jl:748 [inlined]
 [6] *(::Adjoint{Int64,Array{Int64,1}}, ::Diagonal{Int64,Array{Int64,1}}) at /Users/davidsanders/development/julia-dev/usr/share/julia/stdlib/v0.7/LinearAlgebra/src/diagonal.jl:484 (repeats 79998 times)

julia> x' * Diagonal(x) * x
Segmentation fault: 11
```

```
julia> versioninfo()
Julia Version 0.7.0-DEV.4909
Commit 5912b14 (2018-04-19 21:08 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin15.0.0)
  CPU: Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
Environment:
```

### PR DESCRIPTION
**Pull request by andreasnoack**
Also add optimized methods for x'D*y to avoid allocating temporary vector

Fixes JuliaLang/LinearAlgebra.jl#517

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -481,6 +481,10 @@ end
 # dismabiguation methods: * of Diagonal and Adj/Trans AbsVec
 *(A::Diagonal, B::Adjoint{<:Any,<:AbstractVector}) = A * copy(B)
 *(A::Diagonal, B::Transpose{<:Any,<:AbstractVector}) = A * copy(B)
-*(A::Adjoint{<:Any,<:AbstractVector}, B::Diagonal) = copy(A) * B
-*(A::Transpose{<:Any,<:AbstractVector}, B::Diagonal) = copy(A) * B
+*(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal) = Adjoint(map((t,s) -> t'*s, D.diag, parent(x)))
```
Comment (Sacha0) [inline stdlib/LinearAlgebra/src/diagonal.jl]: Perhaps `(D' * x')'` would be simpler / clearer, if methods appropriate for `D' * x'` exist?

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 2]
```diff
@@ -481,6 +481,10 @@ end
...
+*(x::Adjoint{<:Any,<:AbstractVector}, D::Diagonal, y::AbstractVector) =
+    mapreduce(t -> t[1]'*t[2]*t[3], +, zip(parent(x), D.diag, y))
```
Comment (Sacha0) [inline stdlib/LinearAlgebra/src/diagonal.jl]: Alternatively `mapreduce(t -> t[1]*t[2]*t[3], +, zip(x, D.diag, y))`?

**Review by Sacha0 — COMMENTED** [review]
Much thanks for patching these methods Andreas! Please find a few minor comments attached (edited to improve clarity after posting).

**Comment by andreasnoack** [comment]
Looks like this is working now.

**Review by Sacha0 — APPROVED** [review]
Assuming the method deletions do not break an appreciable amount of external code, lgtm! :)

**Pull request merged**